<robot name="spot" xmlns:xacro="http://www.ros.org/wiki/xacro">

  <link name="base_link">
    <visual>
      <origin xyz="0 0 0.68" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://spot_ml_project/models/meshes/base.obj"/>
      </geometry>
      <material name="base_link">
        <color rgba="1 0.8431 0 1"/>
      </material>
    </visual>
    <inertial>
        <mass value="11.44"/>
        <inertia ixx="0.081" ixy="0.0" ixz="-0.001" iyy="0.646" iyz="0.0" izz="0.666"/>
    </inertial>
  </link>

  <gazebo reference="base_link">
    <sensor name="right_camera" type="camera">
          <pose>-0.125 -0.12 0.715 0 0.2 -1.571</pose>
          <camera>
            <horizontal_fov>1.047</horizontal_fov>
            <image>
              <width>800</width>
              <height>600</height>
            </image>
          </camera>
          <always_on>1</always_on>
          <update_rate>30</update_rate>
          <visualize>false</visualize>
          <plugin name="camera_controller" filename="libgazebo_ros_camera.so"> <!-- Camera Image Plugin -->
              <robotNamespace>/</robotNamespace>
              <cameraName>right_camera</cameraName>
              <imageTopicName>image_raw</imageTopicName>
              <cameraInfoTopicName>camera_info</cameraInfoTopicName>
          </plugin>
    </sensor>
  </gazebo>

  <gazebo reference="base_link">
    <sensor name="left_camera" type="camera">
          <pose>-0.125 0.12 0.715 0 0.2 1.571</pose>
          <camera>
            <horizontal_fov>1.047</horizontal_fov>
            <image>
              <width>800</width>
              <height>600</height>
            </image>
          </camera>
          <always_on>1</always_on>
          <update_rate>30</update_rate>
          <visualize>false</visualize>
          <plugin name="camera_controller" filename="libgazebo_ros_camera.so"> <!-- Camera Image Plugin -->
              <robotNamespace>/</robotNamespace>
              <cameraName>left_camera</cameraName>
              <imageTopicName>image_raw</imageTopicName>
              <cameraInfoTopicName>camera_info</cameraInfoTopicName>
          </plugin>
    </sensor>
  </gazebo>

  <gazebo reference="base_link">
    <sensor name="front_right_camera" type="camera">
          <pose>0.415 -0.037 0.67 -1.355558 0.353917 0.554196</pose>
          <camera>
            <horizontal_fov>1.047</horizontal_fov>
            <image>
              <width>800</width>
              <height>600</height>
            </image>
          </camera>
          <always_on>1</always_on>
          <update_rate>30</update_rate>
          <visualize>false</visualize>
          <plugin name="camera_controller" filename="libgazebo_ros_camera.so"> <!-- Camera Image Plugin -->
              <robotNamespace>/</robotNamespace>
              <cameraName>front_right_camera</cameraName>
              <imageTopicName>image_raw</imageTopicName>
              <cameraInfoTopicName>camera_info</cameraInfoTopicName>
          </plugin>
    </sensor>
  </gazebo>

    <gazebo reference="base_link">
    <sensor name="back_camera" type="camera">
          <pose>-0.425 0 0.69 0 0.3 -3.142</pose>
          <camera>
            <horizontal_fov>1.047</horizontal_fov>
            <image>
              <width>800</width>
              <height>600</height>
            </image>
          </camera>
          <always_on>1</always_on>
          <update_rate>30</update_rate>
          <visualize>false</visualize>
          <plugin name="camera_controller" filename="libgazebo_ros_camera.so"> <!-- Camera Image Plugin -->
              <robotNamespace>/</robotNamespace>
              <cameraName>back_camera</cameraName>
              <imageTopicName>image_raw</imageTopicName>
              <cameraInfoTopicName>camera_info</cameraInfoTopicName>
          </plugin>
    </sensor>
  </gazebo>

    <gazebo reference="base_link">
    <sensor name="front_left_camera" type="camera">
          <pose>0.415 0.037 0.67 1.355558 0.353917 -0.554196</pose>
          <camera>
            <horizontal_fov>1.047</horizontal_fov>
            <image>
              <width>800</width>
              <height>600</height>
            </image>
          </camera>
          <always_on>1</always_on>
          <update_rate>30</update_rate>
          <visualize>false</visualize>
          <plugin name="camera_controller" filename="libgazebo_ros_camera.so"> <!-- Camerea Image Plugin -->
              <robotNamespace>/</robotNamespace>
              <cameraName>front_left_camera</cameraName>
              <imageTopicName>image_raw</imageTopicName>
              <cameraInfoTopicName>camera_info</cameraInfoTopicName>
          </plugin>
    </sensor>
  </gazebo>
  
  <link name="front_left.hip">
    <visual>
      <origin xyz="0 0 0.68" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://spot_ml_project/models/meshes/fl.hip.obj"/>
      </geometry>
      <material name="front_left.hip">
        <color rgba="0.2 0.2 0.2 1"/>
      </material>
    </visual>
    <inertial>
        <mass value="2.145"/>
        <inertia ixx="0.001" ixy="0.0" ixz="0.0" iyy="0.001" iyz="0.0" izz="0.001"/>
    </inertial>
  </link>


  <joint name="front_left.hx" type="revolute">
    <origin xyz="0.29785 0.05500 0.00000" rpy="0 0 0"/>
    <axis xyz="1 0 0"/>
    <parent link="base_link"/>
    <child link="front_left.hip"/>
    <limit effort="1000" velocity="1000.00" lower="-0.78539816339744827899" upper="0.78539816339744827899"/>
  </joint>

  <link name="front_left.upperleg">
    <visual>
      <origin xyz="0 0 0.68" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://spot_ml_project/models/meshes/fl.uleg.obj"/>
      </geometry>
      <material name="front_left.upperleg">
        <color rgba="1 0.8431 0 1"/>
      </material>
    </visual>
    <inertial>
        <mass value="1.43"/>
        <inertia ixx="0.017" ixy="0.0" ixz="0.0" iyy="0.017" iyz="0.0" izz="0.002"/>
    </inertial>
  </link>

  <joint name="front_left.hy" type="revolute">
    <origin xyz="0.0 0.11080 0.00" rpy="0 0 0"/>
    <axis xyz="0 1 0"/>
    <parent link="front_left.hip"/>
    <child link="front_left.upperleg"/>
    <limit effort="1000" velocity="1000.00" lower="-0.89884456477707963539" upper="2.2951079663725435509"/>
  </joint>

  <link name="front_left.lowerleg">
    <visual>
      <origin xyz="0 0 0.68" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://spot_ml_project/models/meshes/fr.lleg.obj"/>
      </geometry>
      <material name="front_left.lowerleg">
        <color rgba="0.2 0.2 0.2 1"/>
      </material>
    </visual>
    <inertial>
        <mass value="0.715"/>
        <inertia ixx="0.004" ixy="0.0" ixz="0.0" iyy="0.004" iyz="0.0" izz="0.0"/>
    </inertial>
  </link>

  <joint name="front_left.knee" type="revolute">
    <origin xyz="0.025 0.000 -0.3200" rpy="0 0 0"/>
    <axis xyz="0 1 0"/>
    <parent link="front_left.upperleg"/>
    <child link="front_left.lowerleg"/>
    <limit effort="1000" velocity="1000.00" lower="-2.7929" upper="-0.254402"/>
  </joint>

  <link name="front_right.hip">
    <visual>
      <origin xyz="0 0 0.68" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://spot_ml_project/models/meshes/fr.hip.obj"/>
      </geometry>
      <material name="front_right.hip">
        <color rgba="0.2 0.2 0.2 1"/>
      </material>
    </visual>
    <inertial>
        <mass value="2.145"/>
        <inertia ixx="0.001" ixy="0.0" ixz="0.0" iyy="0.001" iyz="0.0" izz="0.001"/>
    </inertial>
  </link>

  <joint name="front_right.hx" type="revolute">
    <origin xyz="0.29785 -0.05500 0.00000" rpy="0 0 0"/>
    <axis xyz="1 0 0"/>
    <parent link="base_link"/>
    <child link="front_right.hip"/>
    <limit effort="1000" velocity="1000.00" lower="-0.78539816339744827899" upper="0.78539816339744827899"/>
  </joint>

  <link name="front_right.upperleg">
    <visual>
      <origin xyz="0 0 0.68" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://spot_ml_project/models/meshes/fr.uleg.obj"/>
      </geometry>
      <material name="front_right.upperleg">
        <color rgba="1 0.8431 0 1"/>
      </material>
    </visual>
    <inertial>
        <mass value="1.43"/>
        <inertia ixx="0.017" ixy="0.0" ixz="0.0" iyy="0.017" iyz="0.0" izz="0.002"/>
    </inertial>
  </link>

  <joint name="front_right.hy" type="revolute">
    <origin xyz="0.0 -0.11080 0.00" rpy="0 0 0"/>
    <axis xyz="0 1 0"/>
    <parent link="front_right.hip"/>
    <child link="front_right.upperleg"/>
    <limit effort="1000" velocity="1000.00" lower="-0.898845" upper="2.24363"/>
  </joint>

  <link name="front_right.lowerleg">
    <visual>
      <origin xyz="0 0 0.68" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://spot_ml_project/models/meshes/fr.lleg.obj"/>
      </geometry>
      <material name="front_right.lowerleg">
        <color rgba="0.2 0.2 0.2 1"/>
      </material>
    </visual>
    <inertial>
        <mass value="0.715"/>
        <inertia ixx="0.004" ixy="0.0" ixz="0.0" iyy="0.004" iyz="0.0" izz="0.0"/>
    </inertial>
  </link>

  <joint name="front_right.knee" type="revolute">
    <origin xyz="0.025 0.000 -0.3200" rpy="0 0 0"/>
    <axis xyz="0 1 0"/>
    <parent link="front_right.upperleg"/>
    <child link="front_right.lowerleg"/>
    <limit effort="1000" velocity="1000.00" lower="-2.7929" upper="-0.255648"/>
  </joint>

  <link name="hind_left.hip">
    <visual>
      <origin xyz="0 0 0.68" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://spot_ml_project/models/meshes/hl.hip.obj"/>
      </geometry>
      <material name="hl.hip">
        <color rgba="0.2 0.2 0.2 1"/>
      </material>
    </visual>
    <inertial>
        <mass value="2.145"/>
        <inertia ixx="0.001" ixy="0.0" ixz="0.0" iyy="0.001" iyz="0.0" izz="0.001"/>
    </inertial>
  </link>

  <joint name="hind_left.hx" type="revolute">
    <origin xyz="-0.29785 0.05500 0.00" rpy="0 0 0"/>
    <axis xyz="1 0 0"/>
    <parent link="base_link"/>
    <child link="hind_left.hip"/>
    <limit effort="1000" velocity="1000.00" lower="-0.78539816339744827899" upper="0.78539816339744827899"/>
  </joint>

  <link name="hind_left.upperleg">
    <visual>
      <origin xyz="0 0 0.68" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://spot_ml_project/models/meshes/hl.uleg.obj"/>
      </geometry>
      <material name="hind_left.upperleg">
        <color rgba="1 0.8431 0 1"/>
      </material>
    </visual>
    <inertial>
        <mass value="1.43"/>
        <inertia ixx="0.017" ixy="0.0" ixz="0.0" iyy="0.017" iyz="0.0" izz="0.002"/>
    </inertial>
  </link>

  <joint name="hind_left.hy" type="revolute">
    <origin xyz="0.0 0.11080 0.00" rpy="0 0 0"/>
    <axis xyz="0 1 0"/>
    <parent link="hind_left.hip"/>
    <child link="hind_left.upperleg"/>
    <limit effort="1000" velocity="1000.00" lower="-0.89884456477707963539" upper="2.2951079663725435509"/>
  </joint>

  <link name="hind_left.lowerleg">
    <visual>
      <origin xyz="0 0 0.68" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://spot_ml_project/models/meshes/hl.lleg.obj"/>
      </geometry>
      <material name="hind_left.lowerleg">
        <color rgba="0.2 0.2 0.2 1"/>
      </material>
    </visual>
    <inertial>
        <mass value="0.715"/>
        <inertia ixx="0.004" ixy="0.0" ixz="0.0" iyy="0.004" iyz="0.0" izz="0.0"/>
    </inertial>
  </link>

  <joint name="hind_left.knee" type="revolute">
    <origin xyz="0.025 0.000 -0.3200" rpy="0 0 0"/>
    <axis xyz="0 1 0"/>
    <parent link="hind_left.upperleg"/>
    <child link="hind_left.lowerleg"/>
    <limit effort="1000" velocity="1000.00" lower="-2.7929" upper="-0.247067"/>
  </joint>

  <link name="hind_right.hip">
    <visual>
      <origin xyz="0 0 0.68" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://spot_ml_project/models/meshes/hr.hip.obj"/>
      </geometry>
      <material name="hr.hip">
        <color rgba="0.2 0.2 0.2 1"/>
      </material>
    </visual>
    <inertial>
        <mass value="2.145"/>
        <inertia ixx="0.001" ixy="0.0" ixz="0.0" iyy="0.001" iyz="0.0" izz="0.001"/>
    </inertial>
  </link>

  <joint name="hind_right.hx" type="revolute">
    <origin xyz="-0.29785 -0.05500 0.00000" rpy="0 0 0"/>
    <axis xyz="1 0 0"/>
    <parent link="base_link"/>
    <child link="hind_right.hip"/>
    <limit effort="1000" velocity="1000.00" lower="-0.78539816339744827899" upper="0.78539816339744827899"/>
  </joint>

  <link name="hind_right.upperleg">
    <visual>
      <origin xyz="0 0 0.68" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://spot_ml_project/models/meshes/hr.uleg.obj"/>
      </geometry>
      <material name="hind_rightr.upperleg">
        <color rgba="1 0.8431 0 1"/>
      </material>
    </visual>
    <inertial>
        <mass value="1.43"/>
        <inertia ixx="0.017" ixy="0.0" ixz="0.0" iyy="0.017" iyz="0.0" izz="0.002"/>
    </inertial>
  </link>

  <joint name="hind_right.hy" type="revolute">
    <origin xyz="0.0 -0.11080 0.00" rpy="0 0 0"/>
    <axis xyz="0 1 0"/>
    <parent link="hind_right.hip"/>
    <child link="hind_right.upperleg"/>
    <limit effort="1000" velocity="1000.00" lower="-0.89884456477707963539" upper="2.2951079663725435509"/>
  </joint>

  <link name="hind_right.lowerleg">
    <visual>
      <origin xyz="0 0 0.68" rpy="0 0 0"/>
      <geometry>
        <mesh filename="package://spot_ml_project/models/meshes/hr.lleg.obj"/>
      </geometry>
      <material name="hind_right.lowerleg">
        <color rgba="0.2 0.2 0.2 1"/>
      </material>
    </visual>
    <inertial>
        <mass value="0.715"/>
        <inertia ixx="0.004" ixy="0.0" ixz="0.0" iyy="0.004" iyz="0.0" izz="0.0"/>
    </inertial>
  </link>

  <joint name="hind_right.knee" type="revolute">
    <origin xyz="0.025 0.000 -0.3200" rpy="0 0 0"/>
    <axis xyz="0 1 0"/>
    <parent link="hind_right.upperleg"/>
    <child link="hind_right.lowerleg"/>
    <limit effort="1000" velocity="1000.00" lower="-2.7929" upper="-0.248282"/>
  </joint>
  
</robot>
